libsplinter/src/network/peer_manager/connector.rs

```diff
@@ -0,0 +1,167 @@
+// Copyright 2018-2020 Cargill Incorporated
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::mpsc::{channel, Sender};
+
+use super::error::{
+    PeerListError, PeerManagerError, PeerRefAddError, PeerRefRemoveError, PeerRefUpdateError,
+};
+use super::notification::PeerNotificationIter;
+use super::PeerRef;
+use super::{PeerManagerMessage, PeerManagerRequest};
+
+/// The PeerManagerConnector will be used to make requests to the PeerManager.
+///
+/// The connector includes functions to add a new peer reference, update a peer and list the
+/// existing peers.
+#[derive(Clone, Debug)]
+pub struct PeerManagerConnector {
+    sender: Sender<PeerManagerMessage>,
+}
+
+impl PeerManagerConnector {
+    pub(crate) fn new(sender: Sender<PeerManagerMessage>) -> Self {
+        PeerManagerConnector { sender }
+    }
+
+    /// Request that a peer is added to the PeerManager. If a peer already exists, the peer's ref
+    /// count will be incremented
+    ///
+    /// # Arguments
+    ///
+    /// * `peer_id` -  The unique id for the peer.
+    /// * `endpoints` -  The list of endpoints associated with the peer. The list should be in
+    ///     preference order, with the first endpoint being the first attempted.
+    ///
+    /// Returns a PeerRef, that when dropped, will automatically send a removal request to the
+    /// PeerManager.
+    pub fn add_peer_ref(
+        &self,
+        peer_id: String,
+        endpoints: Vec<String>,
+    ) -> Result<PeerRef, PeerRefAddError> {
+        let (sender, recv) = channel();
+
+        let message = PeerManagerMessage::Request(PeerManagerRequest::AddPeer {
+            peer_id,
+            endpoints,
+            sender,
+        });
+
+        match self.sender.send(message) {
+            Ok(()) => (),
+            Err(_) => return Err(PeerRefAddError::Disconnected),
```

Comment:
What does this error tell the caller?  That the peer has disconnected or the that PeerManager's background thread has shutdown?

Reply:
That the PeerManager's background thread has shut down, the display for this error is "peer manager is disconnected"

Reply:
I think my concern with this error variant is that we are also talking about a network concept, so "Disconnected" could apply to that as well.  Perhaps this would be better to have an `InternalError` or `BackgroundThreadMissing` or something with a better naming.  

Ultimately,  It should indicate something that the caller can used to recover (though in this case, the caller probably should panic)

Reply:
I switched it to InternalError with a message saying the PeerManager's receiver has dropped.